# Code Conventions - Phuong Long Vo Dao

## Principes G√©n√©raux

### Code Quality Mantras
1. **Lisibilit√© > Concision** : Code lu 10x plus que √©crit
2. **Explicite > Implicite** : Pas de "magie", comportement clair
3. **Simple > Clever** : Solutions simples et maintenables
4. **DRY** : Don't Repeat Yourself (mais pas avant 3√®me occurrence)
5. **YAGNI** : You Aren't Gonna Need It (pas de code sp√©culatif)
6. **Single Responsibility** : Une fonction/composant = une responsabilit√©

### Philosophie
- **Tests avant refactoring** : S√©curit√© filet avant optimisation
- **Petits commits atomiques** : 1 feature/fix = 1 commit
- **Comments expliquent POURQUOI, pas QUOI** : Code auto-document√©
- **Fail fast** : Validation early, errors explicites

---

## TypeScript Standards

### Configuration
```json
// tsconfig.json (strict mode)
{
  "compilerOptions": {
    "strict": true,
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "noImplicitReturns": true,
    "noFallthroughCasesInSwitch": true
  }
}
```

### Types & Interfaces

**R√®gle 1** : Pr√©f√©rer `interface` pour objets, `type` pour unions/intersections/aliases

```typescript
// ‚úÖ GOOD - Interface pour objets
interface User {
  id: string;
  email: string;
  name: string | null;
}

interface UserWithProfile extends User {
  profile: UserProfile;
}

// ‚úÖ GOOD - Type pour unions
type UserRole = 'admin' | 'moderator' | 'coach' | 'user';
type Status = 'pending' | 'active' | 'suspended';

// ‚úÖ GOOD - Type pour intersection
type AuthenticatedUser = User & { token: string };

// ‚ùå BAD - any interdit
function processData(data: any) { // JAMAIS √ßa
  return data.value;
}

// ‚úÖ GOOD - unknown si vraiment type inconnu
function processData(data: unknown) {
  if (typeof data === 'object' && data !== null && 'value' in data) {
    return (data as { value: string }).value;
  }
  throw new Error('Invalid data structure');
}
```

**R√®gle 2** : √âviter duplication types (cr√©er types partag√©s)

```typescript
// ‚úÖ GOOD - Types dans lib/types/
// lib/types/user.ts
export interface User {
  id: string;
  email: string;
  // ...
}

// lib/types/api.ts
export interface ApiResponse<T> {
  success: boolean;
  data?: T;
  error?: ApiError;
}

// ‚ùå BAD - Duplication
// page1.tsx
interface User { id: string; email: string; }
// page2.tsx
interface User { id: string; email: string; } // Duplication!
```

**R√®gle 3** : Utiliser `as const` pour objets immutables

```typescript
// ‚úÖ GOOD
const USER_ROLES = {
  ADMIN: 'admin',
  MODERATOR: 'moderator',
  COACH: 'coach',
  USER: 'user',
} as const;

type UserRole = typeof USER_ROLES[keyof typeof USER_ROLES];

// ‚ùå BAD
const USER_ROLES = {
  ADMIN: 'admin', // Type: string (trop large)
};
```

### Enums vs Union Types

**Pr√©f√©rer Union Types** (plus l√©ger, meilleur tree-shaking)

```typescript
// ‚úÖ GOOD - Union type
type EventType = 'competition' | 'stage' | 'demonstration' | 'seminar';

// ‚ö†Ô∏è OK mais moins optimal - Enum (uniquement si besoin valeurs num√©riques)
enum EventType {
  Competition = 'competition',
  Stage = 'stage',
}
```

---

## Naming Conventions

### Variables & Fonctions

```typescript
// camelCase pour variables, fonctions
const userProfile = { ... };
const isAuthenticated = true;
const hasPermission = (user: User, action: string) => { ... };

// PascalCase pour composants React, classes, types, interfaces
function UserProfile() { ... }
class DatabaseService { ... }
interface ApiResponse { ... }

// SCREAMING_SNAKE_CASE pour constantes globales
const MAX_UPLOAD_SIZE = 5_000_000; // 5MB
const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL;

// kebab-case pour fichiers (sauf composants React)
// utils/format-date.ts ‚úÖ
// components/UserProfile.tsx ‚úÖ (PascalCase pour composants)
```

### Booleans

**Pr√©fixes** : `is`, `has`, `should`, `can`, `will`

```typescript
// ‚úÖ GOOD
const isLoading = true;
const hasAccess = checkAccess();
const shouldRender = isVisible && isReady;
const canEdit = user.role === 'admin';

// ‚ùå BAD
const loading = true; // Ambigu
const access = checkAccess(); // Pas clair que c'est boolean
```

### Fonctions

**Verbes d'action** : `get`, `set`, `fetch`, `create`, `update`, `delete`, `validate`, `handle`, `toggle`, `calculate`

```typescript
// ‚úÖ GOOD
function getUser(id: string): User { ... }
function fetchClubs(): Promise<Club[]> { ... }
function createBlogPost(data: BlogPostInput): Promise<BlogPost> { ... }
function validateEmail(email: string): boolean { ... }
function handleSubmit(event: FormEvent): void { ... }

// ‚ùå BAD
function user(id: string) { ... } // Pas de verbe
function clubs() { ... } // Ambigu (getter? liste?)
```

### √âv√©nements & Handlers

```typescript
// ‚úÖ GOOD - Handlers pr√©fix√©s "handle" + Event
function handleClick() { ... }
function handleSubmit(event: FormEvent) { ... }
function handleChange(value: string) { ... }

// Props √©v√©nements pr√©fix√©es "on" + Action
interface ButtonProps {
  onClick?: () => void;
  onSubmit?: (data: FormData) => void;
}

// ‚ùå BAD
function click() { ... } // Pas de contexte
function submitForm() { ... } // OK mais moins coh√©rent
```

---

## React / Next.js Standards

### Composants

**Structure type** :

```tsx
// components/common/Button.tsx

// 1. Imports (group√©s logiquement)
import { ComponentProps, forwardRef } from 'react';
import { cn } from '@/lib/utils';

// 2. Types/Interfaces
interface ButtonProps extends ComponentProps<'button'> {
  variant?: 'primary' | 'secondary' | 'ghost';
  size?: 'sm' | 'md' | 'lg';
  isLoading?: boolean;
}

// 3. Constantes internes (si besoin)
const variantStyles = {
  primary: 'bg-primary text-white hover:bg-primary-dark',
  secondary: 'bg-secondary text-white hover:bg-secondary-dark',
  ghost: 'bg-transparent hover:bg-gray-100',
};

// 4. Composant (export named, pas default)
export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  function Button({ 
    variant = 'primary',
    size = 'md',
    isLoading = false,
    children,
    className,
    ...props 
  }, ref) {
    return (
      <button
        ref={ref}
        className={cn(
          'rounded-lg font-medium transition-colors',
          variantStyles[variant],
          isLoading && 'opacity-50 cursor-not-allowed',
          className
        )}
        disabled={isLoading || props.disabled}
        {...props}
      >
        {isLoading ? <Spinner size="sm" /> : children}
      </button>
    );
  }
);

// 5. Export types si r√©utilis√©s
export type { ButtonProps };
```

**R√®gles** :

1. **Export named (pas default)** : `export function Component() { ... }`
   - Meilleur autocompl√©tion
   - Refactoring plus safe
   - Tree-shaking optimal

2. **Props interface d√©di√©e** : `ComponentNameProps`

3. **Props destructuring** : Avec defaults

4. **Ordre props** :
   - Required props en premier
   - Optional props ensuite
   - `className` et `...rest` √† la fin

5. **forwardRef** si composant peut recevoir ref

### Server vs Client Components

```tsx
// ‚úÖ GOOD - Server Component (par d√©faut, pas de "use client")
// app/(marketing)/clubs/page.tsx
import { supabase } from '@/lib/supabase/server';

export default async function ClubsPage() {
  const { data: clubs } = await supabase.from('clubs').select('*');
  
  return (
    <div>
      {clubs?.map(club => <ClubCard key={club.id} club={club} />)}
    </div>
  );
}

// ‚úÖ GOOD - Client Component (interactivit√© requise)
// components/common/Button.tsx
'use client';

import { useState } from 'react';

export function LikeButton() {
  const [liked, setLiked] = useState(false);
  
  return (
    <button onClick={() => setLiked(!liked)}>
      {liked ? '‚ù§Ô∏è' : 'ü§ç'}
    </button>
  );
}

// ‚ùå BAD - "use client" sans raison
'use client'; // Pas besoin si pas de hooks/interactivit√©!

export function StaticCard({ title }: { title: string }) {
  return <div>{title}</div>; // Aucune interactivit√©
}
```

**R√®gle** : Server Component par d√©faut. `"use client"` UNIQUEMENT si :
- Hooks React (`useState`, `useEffect`, etc.)
- Event handlers (`onClick`, `onChange`, etc.)
- Browser APIs (`window`, `localStorage`, etc.)
- Third-party libraries client-only

### Hooks

**R√®gles** :

1. **Prefix `use`** obligatoire
2. **Custom hooks dans `hooks/`**
3. **Un hook = une responsabilit√©**

```typescript
// ‚úÖ GOOD - hooks/useAuth.ts
export function useAuth() {
  const [user, setUser] = useState<User | null>(null);
  const [isLoading, setIsLoading] = useState(true);
  
  useEffect(() => {
    const unsubscribe = supabase.auth.onAuthStateChange((event, session) => {
      setUser(session?.user ?? null);
      setIsLoading(false);
    });
    
    return () => unsubscribe();
  }, []);
  
  return { user, isLoading };
}

// ‚úÖ GOOD - Usage
function MyComponent() {
  const { user, isLoading } = useAuth();
  
  if (isLoading) return <Spinner />;
  if (!user) return <SignInPrompt />;
  
  return <div>Hello {user.email}</div>;
}

// ‚ùå BAD - Logique auth directement dans composant (duplication)
function MyComponent() {
  const [user, setUser] = useState<User | null>(null);
  // ... duplication de la logique auth dans chaque composant
}
```

### Conditional Rendering

```tsx
// ‚úÖ GOOD - Early returns
function UserProfile({ user }: { user: User | null }) {
  if (!user) return <div>Please sign in</div>;
  if (user.suspended) return <div>Account suspended</div>;
  
  return <div>Welcome {user.name}</div>;
}

// ‚úÖ GOOD - Ternaire simple (max 1 niveau)
{isLoading ? <Spinner /> : <Content />}

// ‚ö†Ô∏è AVOID - Ternaire imbriqu√© (complexe)
{isLoading ? <Spinner /> : error ? <Error /> : data ? <Content /> : <Empty />}

// ‚úÖ GOOD - Refactorer ternaires complexes
function ContentState() {
  if (isLoading) return <Spinner />;
  if (error) return <Error error={error} />;
  if (!data) return <Empty />;
  return <Content data={data} />;
}

// ‚úÖ GOOD - Short-circuit pour render conditionnel
{isAdmin && <AdminPanel />}
{items.length > 0 && <ItemsList items={items} />}
```

---

## API Routes Standards

### Structure Endpoint

```typescript
// app/api/clubs/[id]/route.ts
import { NextRequest, NextResponse } from 'next/server';
import { z } from 'zod';
import { supabase } from '@/lib/supabase/server';

// 1. Schemas Zod validation
const clubUpdateSchema = z.object({
  name: z.string().min(3).max(100),
  city: z.string().min(2),
  description: z.string().optional(),
});

// 2. GET Handler
export async function GET(
  request: NextRequest,
  { params }: { params: { id: string } }
) {
  try {
    const { data: club, error } = await supabase
      .from('clubs')
      .select('*')
      .eq('id', params.id)
      .single();
    
    if (error || !club) {
      return NextResponse.json(
        { success: false, error: { code: 'NOT_FOUND', message: 'Club not found' } },
        { status: 404 }
      );
    }
    
    return NextResponse.json({ success: true, data: club });
  } catch (error) {
    console.error('[CLUBS_GET]', error);
    return NextResponse.json(
      { success: false, error: { code: 'INTERNAL_ERROR', message: 'Internal server error' } },
      { status: 500 }
    );
  }
}

// 3. PATCH Handler
export async function PATCH(
  request: NextRequest,
  { params }: { params: { id: string } }
) {
  try {
    // Auth check
    const { data: { session } } = await supabase.auth.getSession();
    if (!session) {
      return NextResponse.json(
        { success: false, error: { code: 'UNAUTHORIZED', message: 'Authentication required' } },
        { status: 401 }
      );
    }
    
    // Parse & validate body
    const body = await request.json();
    const validated = clubUpdateSchema.safeParse(body);
    
    if (!validated.success) {
      return NextResponse.json(
        { 
          success: false, 
          error: { 
            code: 'VALIDATION_ERROR', 
            message: 'Invalid input',
            details: validated.error.errors 
          } 
        },
        { status: 400 }
      );
    }
    
    // Update
    const { data: club, error } = await supabase
      .from('clubs')
      .update(validated.data)
      .eq('id', params.id)
      .select()
      .single();
    
    if (error) {
      return NextResponse.json(
        { success: false, error: { code: 'UPDATE_FAILED', message: error.message } },
        { status: 500 }
      );
    }
    
    return NextResponse.json({ success: true, data: club });
  } catch (error) {
    console.error('[CLUBS_PATCH]', error);
    return NextResponse.json(
      { success: false, error: { code: 'INTERNAL_ERROR', message: 'Internal server error' } },
      { status: 500 }
    );
  }
}
```

**R√®gles** :

1. **Validation Zod** : Toujours valider inputs
2. **Error handling** : try/catch + responses standardis√©es
3. **Auth check** : D√®s le d√©but si endpoint prot√©g√©
4. **Logging** : `console.error('[CONTEXT]', error)` pour debug
5. **HTTP Status codes** : Respecter conventions (200, 201, 400, 401, 403, 404, 500)

### Response Format Standard

```typescript
// Success
{
  "success": true,
  "data": { ... }
}

// Error
{
  "success": false,
  "error": {
    "code": "VALIDATION_ERROR",
    "message": "Human-readable message",
    "details": [ ... ] // Optional (ex: Zod errors)
  }
}
```

---

## Styling Conventions

### Tailwind CSS

**Ordre classes** (recommand√©) :

1. Layout (display, position)
2. Box model (padding, margin, width, height)
3. Typography
4. Visual (colors, borders, shadows)
5. Misc (transitions, transforms)

```tsx
// ‚úÖ GOOD - Ordre logique
<div className="flex items-center justify-between p-4 mb-6 text-lg font-semibold bg-white rounded-lg shadow-md hover:shadow-lg transition-shadow">
  Content
</div>

// ‚ö†Ô∏è OK mais moins lisible - Ordre al√©atoire
<div className="shadow-md mb-6 text-lg rounded-lg p-4 flex font-semibold bg-white items-center transition-shadow hover:shadow-lg justify-between">
  Content
</div>
```

**Responsive** : Mobile-first

```tsx
// ‚úÖ GOOD - Mobile first, puis breakpoints
<div className="text-sm md:text-base lg:text-lg">
  Texte responsive
</div>

<div className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-4">
  {/* Cards */}
</div>
```

**Utiliser `cn()` helper** pour conditions

```typescript
// lib/utils.ts
import { clsx, type ClassValue } from 'clsx';
import { twMerge } from 'tailwind-merge';

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}

// Usage
<button 
  className={cn(
    'px-4 py-2 rounded',
    variant === 'primary' && 'bg-blue-500 text-white',
    variant === 'secondary' && 'bg-gray-500 text-white',
    isDisabled && 'opacity-50 cursor-not-allowed'
  )}
/>
```

---

## File Organization

### Structure

```
src/
‚îú‚îÄ‚îÄ app/                  # Next.js App Router
‚îÇ   ‚îú‚îÄ‚îÄ (marketing)/     # Route group
‚îÇ   ‚îú‚îÄ‚îÄ api/             # API routes
‚îÇ   ‚îî‚îÄ‚îÄ ...
‚îú‚îÄ‚îÄ components/
‚îÇ   ‚îú‚îÄ‚îÄ common/          # R√©utilisables g√©n√©riques
‚îÇ   ‚îú‚îÄ‚îÄ marketing/       # Sp√©cifiques marketing
‚îÇ   ‚îú‚îÄ‚îÄ admin/           # Sp√©cifiques admin
‚îÇ   ‚îî‚îÄ‚îÄ ...
‚îú‚îÄ‚îÄ hooks/               # Custom hooks
‚îÇ   ‚îú‚îÄ‚îÄ useAuth.ts
‚îÇ   ‚îî‚îÄ‚îÄ useClubs.ts
‚îú‚îÄ‚îÄ lib/
‚îÇ   ‚îú‚îÄ‚îÄ supabase/        # Supabase clients
‚îÇ   ‚îú‚îÄ‚îÄ stripe/          # Stripe config
‚îÇ   ‚îú‚îÄ‚îÄ utils/           # Utilities g√©n√©riques
‚îÇ   ‚îÇ   ‚îú‚îÄ‚îÄ cn.ts
‚îÇ   ‚îÇ   ‚îú‚îÄ‚îÄ format.ts
‚îÇ   ‚îÇ   ‚îî‚îÄ‚îÄ validators.ts
‚îÇ   ‚îú‚îÄ‚îÄ constants/       # Constantes app
‚îÇ   ‚îÇ   ‚îú‚îÄ‚îÄ roles.ts
‚îÇ   ‚îÇ   ‚îî‚îÄ‚îÄ config.ts
‚îÇ   ‚îî‚îÄ‚îÄ types/           # Types partag√©s
‚îÇ       ‚îú‚îÄ‚îÄ user.ts
‚îÇ       ‚îú‚îÄ‚îÄ club.ts
‚îÇ       ‚îî‚îÄ‚îÄ api.ts
‚îú‚îÄ‚îÄ styles/
‚îÇ   ‚îú‚îÄ‚îÄ globals.css
‚îÇ   ‚îî‚îÄ‚îÄ variables.css
‚îî‚îÄ‚îÄ middleware.ts
```

### R√®gles Fichiers

1. **Un composant = un fichier** (sauf petits helpers internes)
2. **Colocation** : Fichiers li√©s proche (ex: `Button.tsx` + `Button.test.tsx`)
3. **Index exports** : √âviter (rend navigation difficile)
4. **Max 400 lignes** : Si plus, refactorer en sous-composants

---

## Comments & Documentation

### Quand Commenter

```typescript
// ‚úÖ GOOD - Expliquer POURQUOI (business logic non √©vidente)
// Stripe requires amount in cents, not dollars
const amountCents = amountDollars * 100;

// RLS policies don't apply to service role, so we manually check permissions
if (user.role !== 'admin') {
  throw new Error('Unauthorized');
}

// ‚úÖ GOOD - TODO/FIXME avec contexte
// TODO(username): Implement pagination when dataset grows > 1000 items
// FIXME: Race condition possible if user clicks twice rapidly

// ‚ùå BAD - Commenter le QUOI (code self-explanatory)
// Increment counter
counter++;

// Loop through users
users.forEach(user => { ... });

// ‚ùå BAD - Code comment√© (utiliser git history)
// const oldFunction = () => { ... };
```

### JSDoc (pour utils/helpers)

```typescript
/**
 * Formats a date according to French locale
 * @param date - The date to format
 * @param format - Optional format string (default: 'dd/MM/yyyy')
 * @returns Formatted date string
 * @example
 * formatDate(new Date('2025-11-04')) // '04/11/2025'
 */
export function formatDate(date: Date, format = 'dd/MM/yyyy'): string {
  // ...
}
```

---

## Git Commit Conventions

### Format

```
<type>(<scope>): <subject>

<body>

<footer>
```

### Types

- `feat`: Nouvelle feature
- `fix`: Bug fix
- `docs`: Documentation uniquement
- `style`: Formatting, missing semi colons, etc (pas CSS)
- `refactor`: Refactoring code sans changer comportement
- `perf`: Performance improvement
- `test`: Ajout/modification tests
- `chore`: Maintenance, configs, dependencies

### Exemples

```bash
feat(blog): add comment moderation system

- Admin can approve/reject comments
- Email notification on new comment
- RLS policies for comment visibility

Closes #123

---

fix(auth): prevent race condition on logout

User could trigger multiple logout requests causing
session inconsistency.

---

docs(readme): update setup instructions

---

refactor(components): extract shared Button logic

Moved variant styles to separate config object
for better maintainability.

---

perf(api): add database indexes on frequent queries

Added indexes on:
- blog_posts.author_id
- events.club_id
- orders.user_id

Query time reduced from 800ms to 45ms on /api/blog.
```

### R√®gles

1. **Subject** : 
   - Imperative mood ("add" not "added")
   - Max 72 caract√®res
   - Pas de point final
   - Lowercase

2. **Body** : 
   - Expliquer QUOI et POURQUOI (pas COMMENT)
   - Wrap √† 72 caract√®res

3. **Footer** :
   - Issues r√©f√©renc√©es (Closes #123)
   - Breaking changes (BREAKING CHANGE: ...)

---

## Testing Conventions

### Nommage Tests

```typescript
// ‚úÖ GOOD
describe('Button', () => {
  it('renders children correctly', () => { ... });
  it('calls onClick when clicked', () => { ... });
  it('is disabled when isLoading is true', () => { ... });
  it('throws error when invalid variant provided', () => { ... });
});

// ‚ùå BAD
describe('Button', () => {
  it('works', () => { ... }); // Trop vague
  it('test 1', () => { ... }); // Pas descriptif
});
```

### Structure Test

```typescript
it('updates user profile successfully', async () => {
  // Arrange (Setup)
  const user = createMockUser();
  const newData = { name: 'John Doe' };
  
  // Act (Execute)
  const result = await updateUserProfile(user.id, newData);
  
  // Assert (Verify)
  expect(result.success).toBe(true);
  expect(result.data.name).toBe('John Doe');
});
```

---

## Performance Checklist

- [ ] Images utilisent `next/image`
- [ ] Fonts charg√©es via `next/font`
- [ ] Composants lourds lazy-loaded (`dynamic import`)
- [ ] Lists utilisent `key` unique (pas index si ordre change)
- [ ] Expensive computations m√©moris√©es (`useMemo`)
- [ ] Callbacks m√©moris√©s (`useCallback`) si pass√©s en props
- [ ] API responses cach√©es (React Query `staleTime`)
- [ ] Database queries optimis√©es (indexes, select specific fields)

---

**Version** : 1.0  
**Derni√®re mise √† jour** : 2025-11-04  
**Appliqu√© sur** : Tous fichiers projet

